openedx/core/lib/api/exceptions.py

```diff
@@ -0,0 +1,81 @@
+"""
+Custom exceptions, that allow details to be passed as dict values (which can be
+converted to JSON, like other API responses.
+"""
+
+from rest_framework import exceptions
+
+
+# TODO: Override Throttled, UnsupportedMediaType, ValidationError.  These types require
+# more careful handling of arguments.
+
+
+class _DictAPIException(exceptions.APIException):
+    """
+    Intermediate class to allow exceptions to pass dict detail values.  Use by
+    subclassing this along with another subclass of `exceptions.APIException`.
+    """
+    def __init__(self, detail):
+        if isinstance(detail, dict):
+            self.detail = detail
+        else:
+            super(_DictAPIException, self).__init__(detail)
+
+
+class AuthenticationFailed(exceptions.AuthenticationFailed, _DictAPIException):
+    """
+    Override of DRF's AuthenticationFailed exception to allow dictionary responses.
+    """
+    pass
+
+
+class MethodNotAllowed(exceptions.MethodNotAllowed, _DictAPIException):
+    """
+    Override of DRF's MethodNotAllowed exception to allow dictionary responses.
+    """
+    def __init__(self, method, detail=None):
+        if isinstance(detail, dict):
+            self.detail = detail
+        else:
+            super(MethodNotAllowed, self).__init__(method, detail)
+
+
+class NotAcceptable(exceptions.NotAcceptable, _DictAPIException):
+    """
+    Override of DRF's NotAcceptable exception to allow dictionary responses.
+    """
+
+    def __init__(self, detail=None, available_renderers=None):
+        self.available_renderers = available_renderers
+        if isinstance(detail, dict):
+            self.detail = detail
+        else:
+            super(NotAcceptable, self).__init__(detail, available_renderers)
+
+
+class NotAuthenticated(exceptions.NotAuthenticated, _DictAPIException):
+    """
+    Override of DRF's NotAuthenticated exception to allow dictionary responses.
+    """
+    pass
+
+
+class NotFound(exceptions.NotFound, _DictAPIException):
+    """
+    Override of DRF's NotFound exception to allow dictionary responses.
+    """
+    pass
+
+
+class ParseError(exceptions.ParseError, _DictAPIException):
+    """
+    Override of DRF's ParseError exception to allow dictionary responses.
+    """
+    pass
+
+
+class PermissionDenied(exceptions.PermissionDenied, _DictAPIException):
```

Comment:
Correct.  I wanted to make it easier for future developers to use to create robust error responses.  I was torn, though, because some of the error responses are more complex, so I stopped short of implementing all of them, just to respect time constraints on the sprint.  I can remove the unused ones if you want, but they are tested, and share implementation with code that is both tested and used, so I don't think they're a likely source of code rot, so I'm tempted to leave them in place.